synapse/rest/media/v1/preview_url_resource.py

```diff
@@ -56,6 +57,42 @@
 OG_TAG_NAME_MAXLEN = 50
 OG_TAG_VALUE_MAXLEN = 1000
 
+# A map of globs to API endpoints.
+_oembed_globs = {
+    # Twitter.
+    "https://publish.twitter.com/oembed": [
+        "https://twitter.com/*/status/*",
+        "https://*.twitter.com/*/status/*",
+        "https://twitter.com/*/moments/*",
+        "https://*.twitter.com/*/moments/*",
+        # Include the HTTP versions too.
+        "http://twitter.com/*/status/*",
+        "http://*.twitter.com/*/status/*",
+        "http://twitter.com/*/moments/*",
+        "http://*.twitter.com/*/moments/*",
+    ],
+}
+# Convert the globs to regular expressions.
+_oembed_patterns = {}
+for endpoint, globs in _oembed_globs.items():
+    for glob in globs:
+        pattern = re.compile(re.escape(glob).replace("\\*", ".+"))
```

Comment:
I think this will cause URLs like `https://example.com/.twitter.com/foo/status/` to match. I'm guessing we don't actually want to do that.

Reply:
Huh, yeah it seems that we would want to be a bit more specific there.

Oddly it seems to be what other implementations do (actually they seem to replace them with `.*`, not `.+`, but very similar. Some even strip out the following `.` if it is part of the domain).

I'll think a bit more about this. Thanks!

Reply:
Seems like the easiest way to handle this is to limit the domain name globs to `[a-zA-Z0-9_-]+` which should work for most domain names.

Reply:
See 814842bc286513b27e3c55effaeb2a2930638eed